staging/src/k8s.io/client-go/kubernetes/typed/core/v1/service.go

```diff
@@ -43,197 +40,33 @@ type ServicesGetter interface {
 type ServiceInterface interface {
```

Comment:
It seems like this could be a candidate for generics, too?

Reply:
I didn’t want to change the external interface, so I wanted to keep `ServiceInterface` etc. as-is. We still need to code-gen the interfaces because each one can have a different set of methods, including none (`EvictionInterface`, which doesn’t have any shared methods, just `EvictionExpansion`), and we need a type-specific struct to add expansions.

It might be worth generifying further for the most common cases. Since `Service` uses expansions we can’t move everything to the core client, but we could declare (with bad naming, just is just to show the layout)

```go
type ServiceInterface interface {
    generic.InterfaceWithApply[*v1.Service, *v1.ServiceList, *corev1.ServiceApplyConfiguration]
    ServiceExpansion
}
```

but the template engine doesn’t like that (it wants a function inside the interface). I’m looking into fixing that...

Reply:
As far as the implementations goes, my take on it was that we could use a small number of concrete types covering everything, since what is actually visible is limited by the interface; this avoids having to declare structs with a bunch of implementers. My proposal can be simplified by combining multiple structs to handle lists and apply, which would avoid the diamond inheritance problem in the current version.

Reply:
https://github.com/kubernetes/kubernetes/commit/c1b6293048510472bd04c150c80db959d4e309eb seems like the direction.  I still think we could parameterize `FooList` -> `List[Foo]` and `FooApplyConfiguration` -> `ApplyConfiguration[Node]`.

I don't even know what DeleteCollection does - is it really something we should support, but came along after Service (Service is old) and just never got adapted?

Reply:
`DeleteCollection` for `Service` was removed by #63861 because the server side doesn’t (didn’t?) support it.

I agree re. `ApplyConfiguration[T]` and `List[T]`, I’m looking into that.

Reply:
Deleting it from the client because server doesn't support is one answer.  Making the server support it is another.  I just don't know what that entails :)

Reply:
> Deleting it from the client because server doesn't support is one answer. Making the server support it is another. I just don't know what that entails :)

At best, apiserver only papers over delete collection under the hood because we can't actually perform batch writes (i.e. writing multiple objects in a single transaction to etcd) without breaking the watch protocol.

Reply:
I *still* don't know what it does or what it would mean to support it for Services :)

Reply:
hahaha

Reply:
Seriously, though - is there some deeper implication to supporting it?  It seems like a trivial implementatrion should be available to all resources, no?

Reply:
I mean the apiserver can't support it very literally, it can only pretend to support it and what it would do under the hood would be to process each deletion for each object serially.

Reply:
That's my point - I get that SOME resources might want to parallelize but every resource should have a default handler in serial?

Also, this seems like a weird thing to do server-side

Reply:
> Also, this seems like a weird thing to do server-side

I mean totally, we should support writes to multiple objects in a single transaction (or even transactions in general for Kubernetes) but [when we did this](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/event.go#L25-L44) to our definition of what a watch event is, we kinda cornered ourselves.